### ISSUE TITLE
Shorcuts can't be typed as soon as session is loaded

### ISSUE THREAD

**Issue by fulalas** [lxde/lxqt#1301]
This is a very odd bug: if you type any valid shortcut *right after* LXQt GUI is loaded, it doesn't work and you end with all shortcuts unregistered. It seems that shortcuts should be registered before GUI is shown.

**Comment by stefonarch** [comment]
Same here with arch and recent git packages, found out  thanks to this report that another issue is related and  I never found out why or how to reproduce: my panel get's unresponsive that way, for example taskbar windows do not react on clicks, menu does not open and I have to kill the session - no shortcuts, no runner, no menu.

**Comment by technic** [comment]
I can confirm. If I type shortcut right after session start (probably some components are still loading) the lxqt-globalkeysd seems to hang. And that leads to panel get unresponsive. killing lxqt-globalkeysd with `killall -9` brings panel back to life.

**Comment by stefonarch** [comment]
Any news on this? It happens often to me using shortkeys to early after login.

**Comment by fulalas** [comment]
No news. The current Git master still has this bug :( LXQt probably needs more programmers to help solving some critical bugs and add other useful things.

**Comment by tsujan** [comment]
> LXQt probably needs more programmers

Yes, LXQt definitely needs more active developers. Like all human beings, programmers may have limited times and various life problems. Open-source programmers should do their usual jobs too.

**Comment by agaida** [comment]
Best of all - it will take time to find the reason - i can't help here much because i don't have the problem.
No problem in current installations means also nothing to debug. And that's the main problem here i guess.

**Comment by fulalas** [comment]
@agaida, there's no problem. You can simply try a live boot through a VM of Manjaro LXQt: https://sourceforge.net/projects/manjarolinux-community/files/lxqt/17.0.2/

As soon as GUI is loaded, keep pressing ctrl + alt + t (for example) and you'll see the bug happening. Of course, this is not restricted to Manjaro only. :+1:

**Comment by fulalas** [comment]
@agaida, there's no problem. You can simply try a live boot through a VM of Manjaro LXQt: https://sourceforge.net/projects/manjarolinux-community/files/lxqt/17.0.2/

As soon as GUI is loaded, keep pressing ctrl + alt + t (for example) and you'll see the bug happening. Of course, this is not restricted to Manjaro only. :+1:

**Comment by agaida** [comment]
@fulalas  - nope, i'm not interested in testing live systems - for a reason. If that behaviour is in an installed system and reproducible, i'm interested. The only question if bad thing happends only in live systems is: Which overlay - overlayfs or unionfs - and if overlayfs and a not patched kernel - no further processing :D

**Comment by stefonarch** [comment]
Strange enough I cannot reproduce it with arch (32bit)  packages (which are the same for manjaro I guess).

**Comment by tsujan** [comment]
I can't see it with an updated Manjaro Testing.

**Comment by agaida** [comment]
The reason i asked for live oder installed system - the vanilla kernel has a bug inotify for overlayfs - that prevent successfully the work of lxqt-qtplugin and other things that need notifications about changed inodes etc

**Comment by tsujan** [comment]
I use kernel LTS 4.9.46 (kernel 4.12.X had a bad freezing bug, so I didn't dare to boot with 4.12.10).

**Comment by technic** [comment]
Stop hunting kernel witches guys, bug is in our code :)
LXQt has deadlocks. Let me explain one I've recently found to you.
So lets have a look at https://github.com/lxde/lxqt-globalkeys/blob/master/daemon/core.cpp. We have a `Core::run()` loop that listens for two type of X11 events
1) KeyPress to handle shortcuts
2) Messages coming from `mX11RequestsPipe` such as `KeycodeToString` requests
And there is a mutex `mDataMutex`, that causes deadlock in the following way. When lxqt is loading it calls `changeShortcut(...)` (or similar methods) that hold `mDataMutex` and send a message to pipe.  As stated in this report bug happens when we press key at this moment and in the `Core::run()` thread it goes to KeyPress case and locks in line 1159 at `mDataMutex`. On the other hand `chageShortcut(...)` waits for response from pipe (that will never come since `run()` is paused) and also locks, and consequently never releases the mutex.

P.S. this one is related to https://github.com/lxde/lxqt/issues/992

**Comment by fulalas** [comment]
@technic , thanks for your detailed report. Let's hope it gets fixed soon :)

**Comment by palinek** [comment]
As this is rather hard to reproduce, it would be much of help if anyone able to reproduce would post backtrace of the the frozen lxqt-globalkeysd (with debug symbols installed, of course).
Ex.:
```
~$ ps -ef | egrep lxqt-globalkeysd
palco     1433  1338  0 08:26 ?        00:00:00 lxqt-globalkeysd
palco     3390  2723  0 09:48 pts/6    00:00:00 grep -E lxqt-globalkeysd
~$ gdb
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) attach 1433
Attaching to process 1433
[New LWP 1445]
[New LWP 1452]
[New LWP 1454]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f291a31066d in poll () at ../sysdeps/unix/syscall-template.S:84
84      ../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) thread apply all backtrace

Thread 4 (Thread 0x7f2907184700 (LWP 1454)):
#0  0x00007f291a31066d in poll () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007f291a015150 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
#2  0x00007f291a016ee9 in xcb_wait_for_event () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
#3  0x00007f291c006ce8 in _XReadEvents () from /usr/lib/x86_64-linux-gnu/libX11.so.6
#4  0x00007f291bff5f18 in XNextEvent () from /usr/lib/x86_64-linux-gnu/libX11.so.6
#5  0x00005581d11a6909 in Core::run (this=0x7ffd3a0e0e50) at ./daemon/core.cpp:1149
#6  0x00007f291ac1029d in QThreadPrivate::start (arg=0x7ffd3a0e0e50) at thread/qthread_unix.cpp:368
#7  0x00007f29188ed494 in start_thread (arg=0x7f2907184700) at pthread_create.c:333
#8  0x00007f291a319abf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 3 (Thread 0x7f2907985700 (LWP 1452)):
#0  0x00007f291a31066d in poll () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007f2917780119 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007f291778022c in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007f291ae463ff in QEventDispatcherGlib::processEvents (this=0x7f29000008c0, flags=...) at kernel/qeventdispatcher_glib.cpp:423
#4  0x00007f291adebdba in QEventLoop::exec (this=this@entry=0x7f2907984cd0, flags=..., flags@entry=...) at kernel/qeventloop.cpp:212
#5  0x00007f291ac0b3ca in QThread::exec (this=<optimized out>) at thread/qthread.cpp:515
#6  0x00007f291bb06e45 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5DBus.so.5
#7  0x00007f291ac1029d in QThreadPrivate::start (arg=0x7f291bd79d60) at thread/qthread_unix.cpp:368
#8  0x00007f29188ed494 in start_thread (arg=0x7f2907985700) at pthread_create.c:333
#9  0x00007f291a319abf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 2 (Thread 0x7f290e908700 (LWP 1445)):
#0  0x00007f291a31066d in poll () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007f291a015150 in ?? () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
#2  0x00007f291a016ee9 in xcb_wait_for_event () from /usr/lib/x86_64-linux-gnu/libxcb.so.1
#3  0x00007f2910e6af09 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5
#4  0x00007f291ac1029d in QThreadPrivate::start (arg=0x5581d321d830) at thread/qthread_unix.cpp:368
#5  0x00007f29188ed494 in start_thread (arg=0x7f290e908700) at pthread_create.c:333
#6  0x00007f291a319abf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 1 (Thread 0x7f291c4de8c0 (LWP 1433)):
#0  0x00007f291a31066d in poll () at ../sysdeps/unix/syscall-template.S:84
#1  0x00007f2917780119 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007f291778022c in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007f291ae463ff in QEventDispatcherGlib::processEvents (this=0x5581d326b3d0, flags=...) at kernel/qeventdispatcher_glib.cpp:423
#4  0x00007f291adebdba in QEventLoop::exec (this=this@entry=0x7ffd3a0e0d70, flags=..., flags@entry=...) at kernel/qeventloop.cpp:212
#5  0x00007f291adf4d24 in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1291
#6  0x00005581d11a1ab9 in main (argc=<optimized out>, argv=0x7ffd3a0e1088) at ./daemon/main.cpp:254
(gdb) detach
Detaching from program: /usr/bin/lxqt-globalkeysd, process 1433
(gdb) quit
```

**Comment by edwardtulane** [comment]
Ok, so I am affected by this bug and I am using the latest lxqt git packages under arch (64 bit). I rebuilt globalkeys with `-DCMAKE_BUILD_TYPE=Debug`, had it crash on startup by pressing some shortcuts, and this is the backtrace: https://pastebin.com/vEPccNhv
Can you learn something from it? Or do I need to do something different?

**Comment by palinek** [comment]
> Can you learn something from it? Or do I need to do something different?

There aren't any symbols from the `lxqt-globalkeysd` binary :(

**Comment by technic** [comment]
Anyway even potential deadlock should be fixed.
@palinek did you get my idea? Is there another person who can help with globalkeysd code, because we need to think about program logic fix.

**Comment by palinek** [comment]
> did you get my idea?

Yes, i think you're right (also the @edwardtulane's backtrace seems to confirm it). But as you surely realized, the daemon code is unmaintainable (and that's a very polite term). IMO it will be difficult to find a solution for this deadlock w/o a huge refactoring or introducing new bugs...

We definitely should move to `kglobalaccel` ... with it we also gain the support for wayland for free.

**Comment by technic** [comment]
we can try to use `tryLock` in the `run()` loop line 1159, I hope that will work.

**Comment by palinek** [comment]
> we can try to use tryLock in the run() loop line 1159

...and if that fails, then just store the event into "to_process" list?

**Comment by technic** [comment]
> ...and if that fails, then just store the event into "to_process" list?

yes, can do this way.
Still it can be better to spend this time for kf5 `KGlobalAccel` migration...

**Comment by palinek** [comment]
> Still it can be better to spend this time for kf5 KGlobalAccel migration...

I've done it a time ago for runner -> lxde/lxqt-runner#56

**Comment by edwardtulane** [comment]
> There aren't any symbols from the lxqt-globalkeysd binary :(

Uh, this is embarrassing. I was not aware that `makepkg` always removes the symbols if you don't tell it not to do so. For the sake of completeness, here is the proper backtrace: https://pastebin.com/c3xGBLes

**Comment by technic** [comment]
> here is the proper backtrace

I have the same backtrace on my system.

**Comment by palinek** [comment]
> > here is the proper backtrace

> I have the same backtrace on my system.

...and that's exactly the deadlock you were talking about :(

**Comment by technic** [comment]
Has @kuzmas left the project?

**Comment by palinek** [comment]
> Has kuzmas left the project?

Yes.

**Comment by kuzmas** [comment]
Officially apparently - no. De facto - yes. ;-)

**Comment by palinek** [comment]
@edwardtulane @technic can you, please, test with the lxde/lxqt-globalkeys#55 ?

**Comment by palinek** [comment]
I've just find a way for reproducing this bug... and the lxde/lxqt-globalkeys#55 doesn't solve this issue completely :(

**Comment by technic** [comment]
Yep, your commit still allows deadlock.

**Comment by technic** [comment]
So we can't process KeyPress events from X11 and Dbus requests for configuring shortcuts at the same time (in parallel), because the latter modifies some data structures, that we have to guard with mutex.
The only thing that two threads do in parallel now is polling: XNextEvent and qt event loop for Dbus. Well, linux kernel can poll for two things in one thread. We can integrat X11 part into qt event loop for example, by using qt `nativeEvent` or by getting X11 file descriptor and watch it with `QtSocketNotifier`.

**Comment by technic** [comment]
please test https://github.com/lxde/lxqt-globalkeys/pull/56
